File: Average of Levels in Binary Tree - Leetcode 637/Average of Levels in Binary Tree - Leetcode 637.cpp
#include <vector>
#include <queue>

class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> avgs;
        if (!root) return avgs;

        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            double sum = 0;
            int size = q.size();
            for (int i = 0; i < size; ++i) {
                TreeNode* node = q.front();
                q.pop();
                sum += node->val;

                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
            }
            avgs.push_back(sum / size);
        }

        return avgs;
    }
};

File: Average of Levels in Binary Tree - Leetcode 637/Average of Levels in Binary Tree - Leetcode 637.java
import java.util.*;

public class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> avgs = new ArrayList<>();
        if (root == null) return avgs;

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            double sum = 0;
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                sum += node.val;

                if (node.left != null) {
                    q.add(node.left);
                }
                if (node.right != null) {
                    q.add(node.right);
                }
            }
            avgs.add(sum / size);
        }

        return avgs;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
